code_search/pipeline/index_update_pipeline.py

```diff
@@ -0,0 +1,127 @@
+import kfp.dsl as dsl
+from kubernetes import client as k8s_client
+from typing import Dict
+
+
+def default_gcp_op(name: str, image: str, command: str = None,
+    arguments: str = None, file_inputs: Dict[dsl.PipelineParam, str] = None,
+    file_outputs: Dict[str, str] = None, is_exit_handler=False):
+  """An operator that mounts the default GCP service account to the container.
+
+    The user-gcp-sa secret is created as part of the kubeflow deployment that
+    stores the access token for kubeflow user service account.
+
+    With this service account, the container has a range of GCP APIs to
+    access to. This service account is automatically created as part of the
+    kubeflow deployment.
+
+    For the list of the GCP APIs this service account can access to, check
+    https://github.com/kubeflow/kubeflow/blob/7b0db0d92d65c0746ac52b000cbc290dac7c62b1/deployment/gke/deployment_manager_configs/iam_bindings_template.yaml#L18
+
+    If you want to call the GCP APIs in a different project, grant the kf-user
+    service account access permission.
+  """
+
+  return (
+      dsl.ContainerOp(
+          name,
+          image,
+          command,
+          arguments,
+          file_inputs,
+          file_outputs,
+          is_exit_handler,
+      )
+      .add_volume(
+          k8s_client.V1Volume(
+              name='gcp-credentials',
+              secret=k8s_client.V1SecretVolumeSource(
+                  secret_name='user-gcp-sa'
+              )
+          )
+      )
+      .add_volume_mount(
+          k8s_client.V1VolumeMount(
+              mount_path='/secret/gcp-credentials',
+              name='gcp-credentials',
+          )
+      )
+      .add_env_variable(
+          k8s_client.V1EnvVar(
+              name='GOOGLE_APPLICATION_CREDENTIALS',
+              value='/secret/gcp-credentials/user-gcp-sa.json'
+          )
+      )
+  )
+
+def dataflow_function_embedding_op(project: 'GcpProject', runner: str, target_dataset: str, problem: str,
+                           data_dir: 'GcsUri',saved_model_dir: 'GcsUri',temp_location: 'GcsUri', staging_location: 'GcsUri',
+                           job_name: str, worker_machine_type: str,
+                           num_workers: int, step_name='dataflow_function_embedding'):
+    return default_gcp_op(
+        name = step_name,
+        image = 'gcr.io/yang-codesearch/code-search-dataflow:v20181124-23deb39',
+        command = [
+            'python2',
+            '-m',
+            'code_search.dataflow.cli.create_function_embeddings',
+        ],
+        arguments = [
```

Comment:
Should we open an issue (per comment in PR description) to eventually switch to launching the ksonnet job to submit the Dataflow job?

I think the main advantage of using the ksonnet components is avoiding code duplication. If we reuse the ksonnet components then we know that the pipeline is launching the same jobs that are launched individually.

I think that would be useful for debugging/development because you could just do a one off run of a step by just running the component.

Reply:
done. https://github.com/kubeflow/examples/issues/363